### ISSUE TITLE
Is it possible to inquire nfft's window function?

### ISSUE THREAD

**Issue by mreineck** [NFFT/nfft#35]
I wonder if there is a reliable way to determine the window function for which a given instance of libnfft3.a of libnfft3.so was compiled. This could greatly help detect problems like people setting FG_PSI when the NFFT was compiled for another window function.

If this is not yet part of the API, do you think it would be beneficial to add something like

`const char * nfft_window_function(void);
`
which returns "gaussian", "bspline", "sinc" etc.?

**Comment by tvolkmer** [comment]
As far as I know, the window function cannot be queried via an API with the current code.

@jenskeiner
Maybe we should add an utility function
   `const char * Y(get_window_function_name)(void);`
to `kernel/util/window.c` with values
   `kaiserbessel`
   `gaussian`
   `bspline`
   `sinc`
   `dirac`
depending on the compiled window function?
(These names coincide with the possible configure values for the `--with-window=` flag.)

**Comment by mreineck** [comment]
I'd really appreciate such an addition!

Hmm, given that nfft often comes precompiled as part of a Linux distribution these days, it might be even worth to think about providing all window functions within a single library, under different names, similar to the different prefixes that are already used for different floating point types...

I.e. something like

```
nfft_kb_trafo(...)
nfft_gauss_trafo(...)
nfft_sinc_trafo(...)
```
But this would of course be a huge change, and I have no idea if it is at all feasible...

**Comment by jenskeiner** [comment]
I agree, we should provide a way to get the name of the window function in the API.

In the early stages, it was decided to choose the window function at compile time, so the respective macros can be expanded right away. I'm not sure how much of a true benefit in terms of speed one actually gets from that over an approach that allows to choose the window function at run time.

For now, we should keep it the way it is, but it would be interesting to know if we could change the approach without compromising on efficiency.

**Comment by jenskeiner** [comment]
Let's keep this one open to track the code change.

**Comment by mreineck** [comment]
Thank you!

Thinking about the details again, maybe it's cleaner to return enum values instead of a string, to avoid dealing with string constants.

And using the fact that currently the default cut-off parameter is different for all window funtions, it is possible to use for the moment an external helper function like this:
```

typedef enum { NFFT_WINDOW_KAISER_BESSEL,
               NFFT_WINDOW_SINC_POWER,
               NFFT_WINDOW_B_SPLINE,
               NFFT_WINDOW_GAUSSIAN,
               NFFT_WINDOW_UNKNOWN } nfft_windowtype;

nfft_windowtype nfft_get_windowtype (void){
  // dummy NFFT plan to determine the default cut-off parameter
  NFFT(plan) my_plan;
  int N0=20, M=1;
  NFFT(init_1d)(&my_plan,N0,M);
  int m = my_plan.m;
  NFFT(finalize)(&my_plan);
  switch(m) {
    case 6:
      return NFFT_WINDOW_KAISER_BESSEL;
    case 9:
      return NFFT_WINDOW_SINC_POWER;
    case 11:
      return NFFT_WINDOW_B_SPLINE;
    case 12:
      return NFFT_WINDOW_GAUSSIAN;
  }
  return NFFT_WINDOW_UNKNOWN;
}
```

**Comment by jenskeiner** [comment]
I have merged #37 to dev adding `Y(get_window_name)` which returns a string. As far as the suggested enum goes, we have to check if that would actually translate into a substantial benefit inside the library or for the user.

So far, we're heavily using macros and can resolve case distinctions at compile time inside the library. Don't see any immediate benefit having the enum there.

Can anyone come up with any case where having the enum would be beneficial? Just want to avoid cluttering the interface with yet another function.

**Comment by mreineck** [comment]
There is no substantial gain from using an enum, I think. The added functionality does what I needed, thanks!
